# Salmon Infectomes 2021

<h2>Overview</h2>

<p>This GitHub pages provides additional files, data, and information, supplementary to the Dissertation.</p>

<h2>Supplementary Material</h2>

<h3>1. commands_2021.txt</h3>

<p>Text file containing Unix commands used for analysis, including:</p>

- Indexing and alignment to O. tshawytscha genome (GCA_002872995.1, 2018)
- Filtering of host reads
- 16s rRNA filtering and classification
- Classification and abundance estimation of microbial reads
- AMR gene analysis 

<h3>2. SessionInfo_2021.txt</h3>

<p>Text file containing relevent session information for each of the available R Markdowns (.RMD).</p>

<h3>3. salmon_infectome_data_2021.xlsx</h3>

<p>Compiled Excel document containing input files for R analysis. This includes microbial counts, gene counts, metadata, and lists of immune genes.</p>

<h3>4. DEG_and_moribund_microbes_2021.xlsx</h3>

<p>Compiled Excel document containing relevent output files from analysis in R. This includes DEGs and a list of microbial species on present in moribund samples.</p>

<h3>5. env_abun_2021.Rmd</h3>

<p>R Markdown file containing the R code used to obtain the raw output/results for assessing abundance of ENV and RPS13 transcript abundances.</p>

<h3>6. infectome_analysis_2021.Rmd</h3>

<p>R Markdown file containing the R code used to obtain the raw output/results for analysis of microbial composition of the salmon.</p>

<h3>7. infectome_analysis_2021.Rmd</h3>

<p>R Markdown file containing the R code used to obtain the raw output/results for analysis of AMR gene presence/absence/abundances in the salmon.</p>

<h3>8. host_immune_response_2021.Rmd</h3>

<p>R Markdown file containing the R code used to obtain the raw output/results for analysis immune response in the salmon.</p>
